src/fem/patch_point_values.hh

```diff
@@ -0,0 +1,1230 @@
+    }
+};
+
+/// Defines reinit operations on bulk points.
+struct bulk_reinit {
+	// element operations
+    template<unsigned int dim>
+    static inline void elop_jac(std::vector<ElOp<3>> &operations, FMT_UNUSED IntTableArena &el_table) {
+        // result matrix(spacedim, dim), input matrix(spacedim, dim+1)
+        auto &op = operations[FeBulk::BulkOps::opJac];
+        auto &jac_value = op.result_matrix();
+        const auto &coords_value = operations[ op.input_ops()[0] ].result_matrix();
+        for (unsigned int i=0; i<3; i++)
+            for (unsigned int j=0; j<dim; j++)
+                jac_value(i,j) = coords_value(i,j+1) - coords_value(i,0);
+    }
+    template<unsigned int dim>
+    static inline void elop_inv_jac(std::vector<ElOp<3>> &operations, FMT_UNUSED IntTableArena &el_table) {
+        // result matrix(spacedim, dim), input matrix(spacedim, dim+1)
+        auto &op = operations[FeBulk::BulkOps::opInvJac];
+        auto &inv_jac_value = op.result_matrix();
+        const auto &jac_value = operations[ op.input_ops()[0] ].result_matrix();
+        inv_jac_value = eigen_arena_tools::inverse<3, dim>(jac_value);
+    }
+    template<unsigned int dim>
+    static inline void elop_jac_det(std::vector<ElOp<3>> &operations, FMT_UNUSED IntTableArena &el_table) {
+        // result double, input matrix(spacedim, dim)
+        auto &op = operations[FeBulk::BulkOps::opJacDet];
+        auto &jac_det_value = op.result_matrix();
+        const auto &jac_value = operations[ op.input_ops()[0] ].result_matrix();
+        jac_det_value(0,0) = eigen_arena_tools::determinant<3, dim>(jac_value).abs();
+    }
+
+    // point operations
+    static inline void ptop_coords(FMT_UNUSED std::vector<ElOp<3>> &operations, FMT_UNUSED IntTableArena &el_table) {
+        // Implement
+    }
+    static inline void ptop_weights(std::vector<ElOp<3>> &operations, PatchArena *arena, const std::vector<double> &point_weights) {
+        auto &op = operations[FeBulk::BulkOps::opWeights];
+        op.allocate_result(point_weights.size(), *arena);
+        auto &weights_value = op.result_matrix();
+        for (uint i=0; i<point_weights.size(); ++i)
+            weights_value(0,0)(i) = point_weights[i];
+    }
+    static inline void ptop_JxW(std::vector<ElOp<3>> &operations, FMT_UNUSED IntTableArena &el_table) {
+        auto &op = operations[FeBulk::BulkOps::opJxW];
+        auto &weights_value = operations[ op.input_ops()[0] ].result_matrix();
+        auto &jac_det_value = operations[ op.input_ops()[1] ].result_matrix();
+        ArenaOVec<double> weights_ovec( weights_value(0,0) );
+        ArenaOVec<double> jac_det_ovec( jac_det_value(0,0) );
+        ArenaOVec<double> jxw_ovec = jac_det_ovec * weights_ovec;
+        auto &jxw_value = op.result_matrix();
+        jxw_value(0,0) = jxw_ovec.get_vec();
+    }
+    static inline void ptop_scalar_shape(std::vector<ElOp<3>> &operations,
```

Comment:
Try to prepare 'shape_values' in correct format to avoid copy. Move to precomputed part of arena.